api/src/main/java/jakarta/enterprise/invoke/InvokerBuilder.java

```diff
@@ -0,0 +1,363 @@
+ * where {@code X} and {@code Y} are types:
+ *
+ * <ul>
+ * <li>{@code static X transform(Y value)}</li>
+ * <li>{@code static X transform(Y value, Consumer<Runnable> cleanup)}</li>
+ * <li>{@code X transform()} &ndash; in this case, {@code Y} is the type of the class that
+ * declares the transformer</li>
+ * </ul>
+ *
+ * An input transformer must produce a type that can be consumed by the invokable method.
+ * Specifically: when {@code X} is any-type, it is not type checked during deployment (where
+ * any-type is recursively defined as: the {@code java.lang.Object} class type, or a type
+ * variable that has no bound, or a type variable whose first bound is any-type). Otherwise,
+ * it is a deployment problem if {@code X} is not assignable to the corresponding type in
+ * the declaration of the invokable method (that is the bean class in case of target instance
+ * transformers, or the corresponding parameter type in case of argument transformers).
+ * {@code Y} is not type checked during deployment, so that input transformers may consume
+ * arbitrary types.
+ * TODO this paragraph refers to "assignability", which needs to be defined somewhere!
+ * <p>
+ * When a transformer is registered for given input, it is called before the invokable
+ * method is invoked, and the outcome of the transformer is used in the invocation
+ * instead of the original value passed to the invoker by its caller.
+ * <p>
+ * If the transformer declares the {@code Consumer<Runnable>} parameter, and the execution
+ * of the transformer calls {@code Consumer.accept()} with some {@code Runnable}, it is
+ * guaranteed that the {@code Runnable} will be called after the invokable method is invoked
+ * but before the invoker returns. These {@code Runnable}s are called <em>cleanup tasks</em>.
+ * The order of cleanup task execution is not specified. Passing a {@code null} cleanup task
+ * to the {@code Consumer} is permitted, but has no effect.
+ * <p>
+ * If an input transformation is configured for an input for which a lookup is also configured,
+ * the lookup is performed first and the transformation is applied to the looked up value.
+ * If the looked up bean for some input is {@code @Dependent}, it is guaranteed that all
+ * cleanup tasks registered by a transformer for that input are called before that looked up
+ * {@code @Dependent} bean is destroyed.
+ *
+ * <h2>Output transformations</h2>
+ *
+ * An invokable method has 2 kinds of outputs: the return value and the thrown exception.
+ * These outputs correspond to the return value of {@link Invoker#invoke(Object, Object[]) Invoker.invoke()}
+ * or its thrown exception, respectively.
+ * <p>
+ * Each output can be transformed by a transformer that has one of the following signatures,
+ * where {@code X} and {@code Y} are types:
+ *
+ * <ul>
+ * <li>{@code static X transform(Y value)}</li>
+ * <li>{@code X transform()} &ndash; in this case, {@code Y} is the type of the class that
+ * declares the transformer</li>
```

Comment:
How do you aim to use this with exception transformer? I think this format only applies to return value transformers or am I missing something @Ladicek?

Reply:
I'm not sure why this couldn't apply to exception transformers...? Could you elaborate?

Reply:
We discussed this over chat and the conclusion is that you can technically use this to register something like `Throwable::getMessage` as a transformer.
However, due to another part of the text here:

> it is a deployment problem if Y is not assignable [...] from java.lang.Throwable in case of exception transformers

You are very limited in use cases for that. In almost all cases, an exception transformer will still be a specific static method with `Throwable` param.

Reply:
That is correct, the usefulness of this style for exception transformers is very low, but there's no reason to forbid it.